api/src/main/java/org/apache/iceberg/io/CloseableIterable.java

```diff
@@ -75,6 +75,31 @@ public CloseableIterator<E> iterator() {
     };
   }
 
+  /**
+   * Will run the given runnable when {@link CloseableIterable#close()} has been called.
+   *
+   * @param iterable             The underlying {@link CloseableIterable} to iterate over
+   * @param onCompletionRunnable The runnable to run after the underlying iterable was closed
+   * @param <E>                  The type of der underlying iterable
+   * @return A new {@link CloseableIterable} where the runnable will be executed
+   * as the final step after {@link CloseableIterable#close()} has been called
+   */
+  static <E> CloseableIterable<E> whenComplete(CloseableIterable<E> iterable, Runnable onCompletionRunnable) {
+    Preconditions.checkNotNull(onCompletionRunnable, "Cannot execute a null Runnable after completion");
+    return new CloseableIterable<E>() {
+      @Override
+      public void close() throws IOException {
+        iterable.close();
+        onCompletionRunnable.run();
```

Comment:
Does the `onCompletionRunnable` need to happen in a `try-finally` block in case `iterable.close()` throws?

Might be something you want to add a test for.

Reply:
good catch, that makes definitely sense. Updated and added a test